From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: uRyanxD <familiarodrigues123ro@gmail.com>
Date: Sun, 26 Jul 2026 00:03:22 -0300
Subject: [PATCH] Fix connection throttle issues

Backported from recent versions of Paper; credits to gunjanjaswal and Warrior

diff --git a/src/main/java/net/minecraft/server/HandshakeListener.java b/src/main/java/net/minecraft/server/HandshakeListener.java
index 68041c3e97e9292c7f27dc21ae1c8c2dc744cd4b..8567fe09ae1ca2e9a1e9a8e89b6667e2983e20d8 100644
--- a/src/main/java/net/minecraft/server/HandshakeListener.java
+++ b/src/main/java/net/minecraft/server/HandshakeListener.java
@@ -29,9 +29,10 @@ public class HandshakeListener implements PacketHandshakingInListener {
 
             // CraftBukkit start - Connection throttle
             try {
-                if (!(this.b.channel.localAddress() instanceof io.netty.channel.unix.DomainSocketAddress)) { // PandaSpigot - the connection throttle is useless when you have a Unix domain socket
-                long currentTime = System.currentTimeMillis();
                 long connectionThrottle = MinecraftServer.getServer().server.getConnectionThrottle();
+                if (connectionThrottle > 0 && !(this.b.channel.localAddress() instanceof io.netty.channel.unix.DomainSocketAddress)) { // PandaSpigot - the connection throttle is useless when you have a Unix domain socket, fix the disabled connection throttle still incrementing the list
+                long currentTime = System.currentTimeMillis();
+                // long connectionThrottle = MinecraftServer.getServer().server.getConnectionThrottle(); // PandaSpigot - Moved up
                 InetAddress address = ((java.net.InetSocketAddress) this.b.getSocketAddress()).getAddress();
 
                 synchronized (throttleTracker) {
@@ -52,7 +53,7 @@ public class HandshakeListener implements PacketHandshakingInListener {
                         java.util.Iterator iter = throttleTracker.entrySet().iterator();
                         while (iter.hasNext()) {
                             java.util.Map.Entry<InetAddress, Long> entry = (java.util.Map.Entry) iter.next();
-                            if (entry.getValue() > connectionThrottle) {
+                            if (currentTime - entry.getValue() > connectionThrottle) { // PandaSpigot - Fix connection throttle cleanup wiping the whole tracker
                                 iter.remove();
                             }
                         }
